<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3 Introduction to R | The Researcher’s Guide to Population Genomic</title>
  <meta name="description" content="This book provides an overview of population genomics, discusses relevant statistics, and includes tested tutorials and workflows that can be applied to population genomic datasets." />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="3 Introduction to R | The Researcher’s Guide to Population Genomic" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This book provides an overview of population genomics, discusses relevant statistics, and includes tested tutorials and workflows that can be applied to population genomic datasets." />
  <meta name="github-repo" content="kfarleigh/RGTPG" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3 Introduction to R | The Researcher’s Guide to Population Genomic" />
  
  <meta name="twitter:description" content="This book provides an overview of population genomics, discusses relevant statistics, and includes tested tutorials and workflows that can be applied to population genomic datasets." />
  

<meta name="author" content="Keaka Farleigh" />


<meta name="date" content="2025-08-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="rna-seq-mapping-pipeline.html"/>
<link rel="next" href="data-wrangling-with-r.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-Y75DCMT5TZ"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-Y75DCMT5TZ');
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Researcher's Guide to Population Genomic Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#about-the-author"><i class="fa fa-check"></i>About the Author</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#motivation"><i class="fa fa-check"></i>Motivation</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#acknowledgements"><i class="fa fa-check"></i>Acknowledgements</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#citing-this-book"><i class="fa fa-check"></i>Citing this book</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#citation"><i class="fa fa-check"></i>Citation</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#disclaimer"><i class="fa fa-check"></i>Disclaimer</a></li>
</ul></li>
<li class="part"><span><b>I Techniques</b></span></li>
<li class="chapter" data-level="1" data-path="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html"><a href="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html"><i class="fa fa-check"></i><b>1</b> Double digest restriction site associated DNA sequencing (ddRADseq)</a>
<ul>
<li class="chapter" data-level="1.1" data-path="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html"><a href="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html#purpose"><i class="fa fa-check"></i><b>1.1</b> Purpose</a></li>
<li class="chapter" data-level="1.2" data-path="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html"><a href="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html#overview"><i class="fa fa-check"></i><b>1.2</b> Overview</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html"><a href="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html#extraction"><i class="fa fa-check"></i><b>1.2.1</b> Extraction</a></li>
<li class="chapter" data-level="1.2.2" data-path="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html"><a href="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html#digestion"><i class="fa fa-check"></i><b>1.2.2</b> Digestion</a></li>
<li class="chapter" data-level="1.2.3" data-path="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html"><a href="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html#ligation"><i class="fa fa-check"></i><b>1.2.3</b> Ligation</a></li>
<li class="chapter" data-level="1.2.4" data-path="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html"><a href="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html#size-selection"><i class="fa fa-check"></i><b>1.2.4</b> Size-selection</a></li>
<li class="chapter" data-level="1.2.5" data-path="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html"><a href="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html#polymerase-chain-reaction-pcr"><i class="fa fa-check"></i><b>1.2.5</b> Polymerase chain reaction (PCR)</a></li>
<li class="chapter" data-level="1.2.6" data-path="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html"><a href="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html#sequence"><i class="fa fa-check"></i><b>1.2.6</b> Sequence</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html"><a href="double-digest-restriction-site-associated-dna-sequencing-ddradseq.html#references"><i class="fa fa-check"></i><b>1.3</b> References</a></li>
</ul></li>
<li class="part"><span><b>II Tutorials</b></span></li>
<li class="chapter" data-level="2" data-path="rna-seq-mapping-pipeline.html"><a href="rna-seq-mapping-pipeline.html"><i class="fa fa-check"></i><b>2</b> RNA-seq mapping pipeline</a>
<ul>
<li class="chapter" data-level="2.1" data-path="rna-seq-mapping-pipeline.html"><a href="rna-seq-mapping-pipeline.html#purpose-1"><i class="fa fa-check"></i><b>2.1</b> Purpose</a></li>
<li class="chapter" data-level="2.2" data-path="rna-seq-mapping-pipeline.html"><a href="rna-seq-mapping-pipeline.html#files-required-for-this-pipeline"><i class="fa fa-check"></i><b>2.2</b> Files required for this pipeline</a></li>
<li class="chapter" data-level="2.3" data-path="rna-seq-mapping-pipeline.html"><a href="rna-seq-mapping-pipeline.html#programs-used-in-this-pipeline"><i class="fa fa-check"></i><b>2.3</b> Programs used in this pipeline</a></li>
<li class="chapter" data-level="2.4" data-path="rna-seq-mapping-pipeline.html"><a href="rna-seq-mapping-pipeline.html#notes-on-this-pipeline"><i class="fa fa-check"></i><b>2.4</b> Notes on this pipeline</a></li>
<li class="chapter" data-level="2.5" data-path="rna-seq-mapping-pipeline.html"><a href="rna-seq-mapping-pipeline.html#overview-1"><i class="fa fa-check"></i><b>2.5</b> Overview</a></li>
<li class="chapter" data-level="2.6" data-path="rna-seq-mapping-pipeline.html"><a href="rna-seq-mapping-pipeline.html#quality-control-with-fastqc-and-trimmomatic"><i class="fa fa-check"></i><b>2.6</b> Quality control with FastQC and Trimmomatic</a></li>
<li class="chapter" data-level="2.7" data-path="rna-seq-mapping-pipeline.html"><a href="rna-seq-mapping-pipeline.html#mapping-with-star"><i class="fa fa-check"></i><b>2.7</b> Mapping with STAR</a></li>
<li class="chapter" data-level="2.8" data-path="rna-seq-mapping-pipeline.html"><a href="rna-seq-mapping-pipeline.html#variant-calling"><i class="fa fa-check"></i><b>2.8</b> Variant calling</a>
<ul>
<li class="chapter" data-level="2.8.1" data-path="rna-seq-mapping-pipeline.html"><a href="rna-seq-mapping-pipeline.html#variant-calling-steps"><i class="fa fa-check"></i><b>2.8.1</b> Variant calling steps</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="rna-seq-mapping-pipeline.html"><a href="rna-seq-mapping-pipeline.html#references-1"><i class="fa fa-check"></i><b>2.9</b> References</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="introduction-to-r.html"><a href="introduction-to-r.html"><i class="fa fa-check"></i><b>3</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="3.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#purpose-2"><i class="fa fa-check"></i><b>3.1</b> Purpose</a></li>
<li class="chapter" data-level="3.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#files-required-for-this-pipeline-1"><i class="fa fa-check"></i><b>3.2</b> Files required for this pipeline</a></li>
<li class="chapter" data-level="3.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#programs-used-in-this-pipeline-1"><i class="fa fa-check"></i><b>3.3</b> Programs used in this pipeline</a></li>
<li class="chapter" data-level="3.4" data-path="introduction-to-r.html"><a href="introduction-to-r.html#notes-on-this-tutorial"><i class="fa fa-check"></i><b>3.4</b> Notes on this tutorial</a></li>
<li class="chapter" data-level="3.5" data-path="introduction-to-r.html"><a href="introduction-to-r.html#overview-2"><i class="fa fa-check"></i><b>3.5</b> Overview</a></li>
<li class="chapter" data-level="3.6" data-path="introduction-to-r.html"><a href="introduction-to-r.html#before-we-begin"><i class="fa fa-check"></i><b>3.6</b> Before we begin</a></li>
<li class="chapter" data-level="3.7" data-path="introduction-to-r.html"><a href="introduction-to-r.html#what-is-r"><i class="fa fa-check"></i><b>3.7</b> What is R?</a></li>
<li class="chapter" data-level="3.8" data-path="introduction-to-r.html"><a href="introduction-to-r.html#advantages-and-disadvantges-of-r"><i class="fa fa-check"></i><b>3.8</b> Advantages and disadvantges of R</a></li>
<li class="chapter" data-level="3.9" data-path="introduction-to-r.html"><a href="introduction-to-r.html#what-is-rstudio"><i class="fa fa-check"></i><b>3.9</b> What is Rstudio?</a></li>
<li class="chapter" data-level="3.10" data-path="introduction-to-r.html"><a href="introduction-to-r.html#the-rstudio-layout"><i class="fa fa-check"></i><b>3.10</b> The Rstudio layout</a>
<ul>
<li class="chapter" data-level="3.10.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#r-console"><i class="fa fa-check"></i><b>3.10.1</b> R console</a></li>
<li class="chapter" data-level="3.10.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#r-script"><i class="fa fa-check"></i><b>3.10.2</b> R script</a></li>
<li class="chapter" data-level="3.10.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#environment"><i class="fa fa-check"></i><b>3.10.3</b> Environment</a></li>
<li class="chapter" data-level="3.10.4" data-path="introduction-to-r.html"><a href="introduction-to-r.html#plots-packages-and-help"><i class="fa fa-check"></i><b>3.10.4</b> Plots, Packages, and Help</a></li>
</ul></li>
<li class="chapter" data-level="3.11" data-path="introduction-to-r.html"><a href="introduction-to-r.html#working-directories"><i class="fa fa-check"></i><b>3.11</b> Working directories</a></li>
<li class="chapter" data-level="3.12" data-path="introduction-to-r.html"><a href="introduction-to-r.html#software-e.g.-packages"><i class="fa fa-check"></i><b>3.12</b> Software (e.g., Packages)</a></li>
<li class="chapter" data-level="3.13" data-path="introduction-to-r.html"><a href="introduction-to-r.html#assigning-variables"><i class="fa fa-check"></i><b>3.13</b> Assigning variables</a></li>
<li class="chapter" data-level="3.14" data-path="introduction-to-r.html"><a href="introduction-to-r.html#types-of-variables-and-data"><i class="fa fa-check"></i><b>3.14</b> Types of variables and data</a>
<ul>
<li class="chapter" data-level="3.14.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>3.14.1</b> Vectors</a></li>
<li class="chapter" data-level="3.14.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-frames"><i class="fa fa-check"></i><b>3.14.2</b> Data frames</a></li>
</ul></li>
<li class="chapter" data-level="3.15" data-path="introduction-to-r.html"><a href="introduction-to-r.html#references-2"><i class="fa fa-check"></i><b>3.15</b> References</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-wrangling-with-r.html"><a href="data-wrangling-with-r.html"><i class="fa fa-check"></i><b>4</b> Data wrangling with R</a>
<ul>
<li class="chapter" data-level="4.1" data-path="data-wrangling-with-r.html"><a href="data-wrangling-with-r.html#purpose-3"><i class="fa fa-check"></i><b>4.1</b> Purpose</a></li>
<li class="chapter" data-level="4.2" data-path="data-wrangling-with-r.html"><a href="data-wrangling-with-r.html#files-required-for-this-pipeline-2"><i class="fa fa-check"></i><b>4.2</b> Files required for this pipeline</a></li>
<li class="chapter" data-level="4.3" data-path="data-wrangling-with-r.html"><a href="data-wrangling-with-r.html#programs-used-in-this-pipeline-2"><i class="fa fa-check"></i><b>4.3</b> Programs used in this pipeline</a></li>
<li class="chapter" data-level="4.4" data-path="data-wrangling-with-r.html"><a href="data-wrangling-with-r.html#notes-on-this-tutorial-1"><i class="fa fa-check"></i><b>4.4</b> Notes on this tutorial</a></li>
<li class="chapter" data-level="4.5" data-path="data-wrangling-with-r.html"><a href="data-wrangling-with-r.html#overview-3"><i class="fa fa-check"></i><b>4.5</b> Overview</a></li>
<li class="chapter" data-level="4.6" data-path="data-wrangling-with-r.html"><a href="data-wrangling-with-r.html#what-is-data-wrangling"><i class="fa fa-check"></i><b>4.6</b> What is data wrangling?</a></li>
<li class="chapter" data-level="4.7" data-path="data-wrangling-with-r.html"><a href="data-wrangling-with-r.html#references-3"><i class="fa fa-check"></i><b>4.7</b> References</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-analysis-and-visualization-in-r.html"><a href="data-analysis-and-visualization-in-r.html"><i class="fa fa-check"></i><b>5</b> Data analysis and visualization in R</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-analysis-and-visualization-in-r.html"><a href="data-analysis-and-visualization-in-r.html#purpose-4"><i class="fa fa-check"></i><b>5.1</b> Purpose</a></li>
<li class="chapter" data-level="5.2" data-path="data-analysis-and-visualization-in-r.html"><a href="data-analysis-and-visualization-in-r.html#files-required-for-this-pipeline-3"><i class="fa fa-check"></i><b>5.2</b> Files required for this pipeline</a></li>
<li class="chapter" data-level="5.3" data-path="data-analysis-and-visualization-in-r.html"><a href="data-analysis-and-visualization-in-r.html#programs-used-in-this-pipeline-3"><i class="fa fa-check"></i><b>5.3</b> Programs used in this pipeline</a></li>
<li class="chapter" data-level="5.4" data-path="data-analysis-and-visualization-in-r.html"><a href="data-analysis-and-visualization-in-r.html#notes-on-this-tutorial-2"><i class="fa fa-check"></i><b>5.4</b> Notes on this tutorial</a></li>
<li class="chapter" data-level="5.5" data-path="data-analysis-and-visualization-in-r.html"><a href="data-analysis-and-visualization-in-r.html#overview-4"><i class="fa fa-check"></i><b>5.5</b> Overview</a></li>
<li class="chapter" data-level="5.6" data-path="data-analysis-and-visualization-in-r.html"><a href="data-analysis-and-visualization-in-r.html#references-4"><i class="fa fa-check"></i><b>5.6</b> References</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">The Researcher’s Guide to Population Genomic</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-r" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">3</span> Introduction to R<a href="introduction-to-r.html#introduction-to-r" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p><em>Written by Keaka Farleigh on August 5th, 2025: edited from materials originally developed by Dr. Tereza Jezkova and Alfredo Ascanio</em></p>
<p><em>Last updated by Keaka Farleigh on August 11th, 2025</em></p>
<div id="purpose-2" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Purpose<a href="introduction-to-r.html#purpose-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This tutorial will provide an introduction to the programming language <a href="https://www.r-project.org/">R</a>. I will assume that you have no experience with R and that you are at the beginning of your journey towards becoming an R expert.</p>
</div>
<div id="files-required-for-this-pipeline-1" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Files required for this pipeline<a href="introduction-to-r.html#files-required-for-this-pipeline-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>None!</p>
</div>
<div id="programs-used-in-this-pipeline-1" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Programs used in this pipeline<a href="introduction-to-r.html#programs-used-in-this-pipeline-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li><a href="https://www.r-project.org/">R</a> (R core team, 2025)</li>
<li><a href="https://posit.co/download/rstudio-desktop/">Rstudio</a></li>
</ul>
</div>
<div id="notes-on-this-tutorial" class="section level2 hasAnchor" number="3.4">
<h2><span class="header-section-number">3.4</span> Notes on this tutorial<a href="introduction-to-r.html#notes-on-this-tutorial" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The materials in this tutorial were originally developed as part of the <a href="https://sites.miamioh.edu/lambda/">Lambda bioinformatic workshop</a> at Miami University by Dr. Tereza Jezkova and Alfredo Ascanio. These materials are now presented at Lambda and the Foundations in R workshop at the University of Virginia.</p>
</div>
<div id="overview-2" class="section level2 hasAnchor" number="3.5">
<h2><span class="header-section-number">3.5</span> Overview<a href="introduction-to-r.html#overview-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We will cover the basics of R and Rstudio before getting some hands-on experience!</p>
</div>
<div id="before-we-begin" class="section level2 hasAnchor" number="3.6">
<h2><span class="header-section-number">3.6</span> Before we begin<a href="introduction-to-r.html#before-we-begin" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Before we get started, I want to set some expectations. Please give yourself some grace, wherever you are in your journey to becoming an R expert. You are learning a <em><strong>new</strong></em> language. There will be failures and you will get error messages you don’t understand, but over time it gets easier as you become more fluent. Please do not be discouraged by the failures or errors, see them as a learning opportunity and please reach out if you need any help, ever.</p>
</div>
<div id="what-is-r" class="section level2 hasAnchor" number="3.7">
<h2><span class="header-section-number">3.7</span> What is R?<a href="introduction-to-r.html#what-is-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>R is a free and open-source (public) programming language that was developed in the 1990’s and published in 2000. It is based on the programming language S, which was developed by Bell labs in the 1970’s.</p>
<p>I list some basic facts about R below:</p>
<ul>
<li><p>R can be run on any platform (e.g., Linux, Mac, Windows).</p></li>
<li><p>R has public repositories that hold software that can be downloaded to perform a variety of analyses. These repositories are maintained by volunteers and require software to be documented and adhere to specific rules to be held in the repository.</p>
<ul>
<li>The official repository is the <a href="https://cran.r-project.org/">CRAN</a>.</li>
<li>The other major repository is <a href="https://www.bioconductor.org/">BioConductor</a>.</li>
</ul></li>
<li><p>R software can also be hosted and downloaded from other repositories such as <a href="https://github.com/">GitHub</a>, but these repositories do not require specific documentation or require developers to adhere to specific practices (some still do).</p></li>
<li><p>There is (and has been) a large community of R users, so there are a lot of helpful resources online if you have questions or run into any errors.</p></li>
</ul>
</div>
<div id="advantages-and-disadvantges-of-r" class="section level2 hasAnchor" number="3.8">
<h2><span class="header-section-number">3.8</span> Advantages and disadvantges of R<a href="introduction-to-r.html#advantages-and-disadvantges-of-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Advantages</p>
<ul>
<li>R is the most popular statistical programming language.</li>
<li>R is open-source and free.</li>
<li>R has many help forums online.</li>
<li>R has an enormous amount of software available to users (currently ~20,000 and growing)</li>
<li>Available software can be used for things like data mining, machine learning, data visualization, and more! You can also develop and publish software if what you need isn’t in the R universe.</li>
</ul>
<p>Disadvantages</p>
<ul>
<li>R has a steep learning curve (it is a whole new language after all).</li>
<li>Different analyses and algorithms are spread across software.</li>
<li>R is memory intensive because it stores all objects in physical memory.</li>
<li>R can be slower then other langauges.</li>
</ul>
</div>
<div id="what-is-rstudio" class="section level2 hasAnchor" number="3.9">
<h2><span class="header-section-number">3.9</span> What is Rstudio?<a href="introduction-to-r.html#what-is-rstudio" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Rstudio is an integrated development environment (IDE) for R. Rstudio is also free and open-source. So what is an IDE? An IDE is a software application that allows you to run analyses, develop software, debug code, and more all in a single interface. Rstudio is often used to run R, because, well, R was developed in the 1990’s and it looks like it. Rstudio allows us to have a more streamlined interface, see everything that we are working on, actively edit our code, and visualize plots as we go. It is more difficult to do these things in base R.</p>
</div>
<div id="the-rstudio-layout" class="section level2 hasAnchor" number="3.10">
<h2><span class="header-section-number">3.10</span> The Rstudio layout<a href="introduction-to-r.html#the-rstudio-layout" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The Rstudio layout has 4 panes or windows. We will walk through each of them.</p>
<p><img src="imgs/Intro2R_Rstudio_window.png" width="2550" style="display: block; margin: auto;" /></p>
<div id="r-console" class="section level3 hasAnchor" number="3.10.1">
<h3><span class="header-section-number">3.10.1</span> R console<a href="introduction-to-r.html#r-console" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The R console is where R is run and results will show up unless you assign them to a variable (more on that a little later). The <code>&gt;</code> indicates that R is ready for a command, when something is running you will see the <code>&gt;</code> go away and a little stop sign by the broom in the top right of the R console will appear. Sometimes, a <code>+</code> sign appears. This mean that your command was missing something. Often times, its a missing parentheses or comma.</p>
<p>Try the code below to see what I am talking about (copy and paste the code into your R console, then hit enter). Can you see why the second line of code didn’t run?</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="introduction-to-r.html#cb15-1" tabindex="-1"></a>working_code <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>)</span>
<span id="cb15-2"><a href="introduction-to-r.html#cb15-2" tabindex="-1"></a></span>
<span id="cb15-3"><a href="introduction-to-r.html#cb15-3" tabindex="-1"></a>plus_sign <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span></span></code></pre></div>
</div>
<div id="r-script" class="section level3 hasAnchor" number="3.10.2">
<h3><span class="header-section-number">3.10.2</span> R script<a href="introduction-to-r.html#r-script" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The R script pane is where you write and edit code. The script you write here will be color coded to make it easier to identify different elements of your code, like functions, comments, colors, and number. Parentheses and quotations will be autofilled for you.You can also comment your code by starting the line with a <code>#</code>. R will recognize this and will not run this line if you execute your entire script at once. Any scripts saved will have a .R file exstension by default (e.g., myscript.R).</p>
<p>Here is an example of code with a comment:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="introduction-to-r.html#cb16-1" tabindex="-1"></a><span class="co"># Look, I made a comment</span></span>
<span id="cb16-2"><a href="introduction-to-r.html#cb16-2" tabindex="-1"></a>working_code <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>)</span></code></pre></div>
<p>So, how do you run code, you ask?</p>
<p>If your cursor is in Rscript window you can press run at the top of that window. Alternatively, you can press <code>Ctrl+Enter</code> or <code>Command+Return</code> on a PC or Mac, respectively.</p>
<p>You can also R code in the R console by typing code after the <code>&gt;</code> and then pressing enter.</p>
</div>
<div id="environment" class="section level3 hasAnchor" number="3.10.3">
<h3><span class="header-section-number">3.10.3</span> Environment<a href="introduction-to-r.html#environment" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The Environment window shows you which variables you have created so far and tells you about them. For example, you can see what size they are, what type they are, and sometimes what is inside of them. If you ran the <code>working_code &lt;- c(1,2,3,4)</code> line above, you should be able to see a variable called <code>working_code</code> that is composed of 4 numbers (1-4). You can also see how much memory is being used by your R session and import files. Importantly, you can save your R workspace, which is all of your objects that you have created in your environment. Think of it as saving a document that you want to pick up where you left off immediately the next time you open it. Saving an R workspace will produce a file with the .Rdata file extension.</p>
</div>
<div id="plots-packages-and-help" class="section level3 hasAnchor" number="3.10.4">
<h3><span class="header-section-number">3.10.4</span> Plots, Packages, and Help<a href="introduction-to-r.html#plots-packages-and-help" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The plots, packages, and help window has a lot. Any plots that you generate will show up here. You can also see what software you have installed (Packages), check the files you have available (Files), and get help if you have a question about a function or package.</p>
<p>Try the code below to get help with the <code>mean</code> function:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="introduction-to-r.html#cb17-1" tabindex="-1"></a>?mean</span></code></pre></div>
</div>
</div>
<div id="working-directories" class="section level2 hasAnchor" number="3.11">
<h2><span class="header-section-number">3.11</span> Working directories<a href="introduction-to-r.html#working-directories" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Like other coding languages, R works from within folders, or directories. Think of this as the location where you are doing your work. In this “working directory” you have immediate access to the files in that location, but if you need files from another location you need to either change into that other directory or use a file path to direct R to that file.</p>
<p>Sometimes, you will get an error saying that a file does not exist, even though you can see it in a folder. This is often because you are in the wrong working directory.</p>
<p>Here are a few ways to check and set your working directory:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="introduction-to-r.html#cb18-1" tabindex="-1"></a><span class="co"># View your working directory</span></span>
<span id="cb18-2"><a href="introduction-to-r.html#cb18-2" tabindex="-1"></a><span class="fu">getwd</span>()</span>
<span id="cb18-3"><a href="introduction-to-r.html#cb18-3" tabindex="-1"></a></span>
<span id="cb18-4"><a href="introduction-to-r.html#cb18-4" tabindex="-1"></a><span class="co"># Set your working directory</span></span>
<span id="cb18-5"><a href="introduction-to-r.html#cb18-5" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&quot;/path/to/your/folder&quot;</span>)</span></code></pre></div>
<p>If you are in Rstudio, you can also use your mouse/touch pad to set your working directory.</p>
<p>Session –&gt; Set Working Directory –&gt; Choose Directory</p>
</div>
<div id="software-e.g.-packages" class="section level2 hasAnchor" number="3.12">
<h2><span class="header-section-number">3.12</span> Software (e.g., Packages)<a href="introduction-to-r.html#software-e.g.-packages" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The software that is available with R are called packages. These packages are bundles of code, data, and documentation that are compiled so that other people can use them. You have to install these packages the first time that you use them, and you have to load them in every R session that you wish to use them.</p>
<p>Here is an example of how to install and load the package <a href="https://dplyr.tidyverse.org/">dplyr</a>:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="introduction-to-r.html#cb19-1" tabindex="-1"></a><span class="co"># Install dplyr</span></span>
<span id="cb19-2"><a href="introduction-to-r.html#cb19-2" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;dplyr&quot;</span>)</span>
<span id="cb19-3"><a href="introduction-to-r.html#cb19-3" tabindex="-1"></a></span>
<span id="cb19-4"><a href="introduction-to-r.html#cb19-4" tabindex="-1"></a><span class="co"># Load dplyr</span></span>
<span id="cb19-5"><a href="introduction-to-r.html#cb19-5" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code></pre></div>
</div>
<div id="assigning-variables" class="section level2 hasAnchor" number="3.13">
<h2><span class="header-section-number">3.13</span> Assigning variables<a href="introduction-to-r.html#assigning-variables" class="anchor-section" aria-label="Anchor link to header"></a></h2>
</div>
<div id="types-of-variables-and-data" class="section level2 hasAnchor" number="3.14">
<h2><span class="header-section-number">3.14</span> Types of variables and data<a href="introduction-to-r.html#types-of-variables-and-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="vectors" class="section level3 hasAnchor" number="3.14.1">
<h3><span class="header-section-number">3.14.1</span> Vectors<a href="introduction-to-r.html#vectors" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="data-frames" class="section level3 hasAnchor" number="3.14.2">
<h3><span class="header-section-number">3.14.2</span> Data frames<a href="introduction-to-r.html#data-frames" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
</div>
<div id="references-2" class="section level2 hasAnchor" number="3.15">
<h2><span class="header-section-number">3.15</span> References<a href="introduction-to-r.html#references-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="rna-seq-mapping-pipeline.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="data-wrangling-with-r.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "day",
    "family": "sans",
    "size": 2,
    "color": "black",
    "download": ["pdf", "epub"]
  },
  "edit": null,
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": null,
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "section",
    "scroll_highlight": true
  },
  "toolbar": {
    "position": "fixed"
  }
});
});
</script>

</body>

</html>
